secboot/encrypt_sb.go

```diff
@@ -56,8 +60,34 @@ func FormatEncryptedDevice(key EncryptionKey, label, node string) error {
 // AddRecoveryKey adds a fallback recovery key rkey to the existing encrypted
 // volume created with FormatEncryptedDevice on the block device given by node.
 // The existing key to the encrypted volume is provided in the key argument.
+//
+// A heuristic memory cost is used.
 func AddRecoveryKey(key EncryptionKey, rkey RecoveryKey, node string) error {
-	return sbAddRecoveryKeyToLUKS2Container(node, key[:], sb.RecoveryKey(rkey), nil)
+	usableMem, err := osutil.TotalUsableMemory()
+	if err != nil {
+		return fmt.Errorf("cannot get usable memory for KDF parameters when adding the recovery key: %v", err)
+	}
+	// The KDF memory is heuristically calculated by taking the
+	// usable memory and subtracting hardcoded 384MB that is
+	// needed to keep the system working. Half of that is the mem
+	// we want to use for the KDF. Doing it this way avoids the expensive
+	// benchmark from cryptsetup. The recovery key is already 128bit
+	// strong so we don't need to be super precise here.
+	kdfMem := (int(usableMem) - 384*1024*1024) / 2
```

Comment:
do we know how much memory we actually have free on the device during snap-bootstrap and during install mode? 384 is kind of for the full user space run mode? just wondering if it's a bit if it's too pessimistic for here.

Reply:
as I said, I don't think it's a blocker but we might want to explore a bit more what to do on master about this